src/framework/components/esmscript/attribute-utils.js

```diff
@@ -0,0 +1,364 @@
+import { Color } from "../../../core/math/color.js";
+import { Curve } from "../../../core/math/curve.js";
+import { CurveSet } from "../../../core/math/curve-set.js";
+import { Vec2 } from "../../../core/math/vec2.js";
+import { Vec3 } from "../../../core/math/vec3.js";
+import { Vec4 } from "../../../core/math/vec4.js";
+import { Asset } from "../../../framework/asset/asset.js";
+import { GraphNode } from "../../../scene/graph-node.js";
+import { Debug } from "../../../core/debug.js";
+
+/**
+ * @callback UpdateFunction
+ * @param {number} dt - The time since the last update.
+ * @ignore
+ */
+
+/**
+ * @callback SwapFunction
+ * @param {Object} newState - The new state to swap to.
+ * @ignore
+ */
+
+/**
+ * This type represents a generic class constructor.
+ * @typedef {Function} ModuleClass
+ */
+
+/**
+ * @typedef {Object} AttributeDefinition
+ * @property {'asset'|'boolean'|'curve'|'entity'|'json'|'number'|'rgb'|'rgba'|'string'|'vec2'|'vec3'|'vec4'} type - The attribute type
+ */
+
+/**
+ * The expected output of an ESM Script file. It contains the class definition and the attributes it requires.
+ * @typedef {Object} ModuleExport
+ * @property {ModuleClass} default - The default export of a esm script that defines a class
+ * @property {Object<string, AttributeDefinition>} attributes - An object containing the names of attributes and their definitions;
+ */
+
+/**
+ * A list of valid attribute types
+ * @ignore
+ */
+export const VALID_ATTR_TYPES = new Set([
+    "asset",
+    "boolean",
+    "curve",
+    "entity",
+    "json",
+    "number",
+    "rgb",
+    "rgba",
+    "string",
+    "vec2",
+    "vec3",
+    "vec4"
+]);
+
+/**
+ * For any given attribute definition returns whether it conforms to the required
+ * shape of an attribute definition.
+ *
+ * @param {AttributeDefinition|object} attributeDefinition - The attribute to check
+ * @returns {boolean} True if the object can be treated as a attribute definition
+ * @example
+ * isValidAttributeDefinition({ type: 'entity' }); // true
+ * isValidAttributeDefinition({ type: 'invalidType' }); // false
+ * isValidAttributeDefinition({ x: 'y' }); // false
+ */
+export const isValidAttributeDefinition = (attributeDefinition) => {
```

Comment:
same here and possibly others, no need to export it?

Reply:
ok, in that case those will need to be at some point reexported in the index.js, and those would be visible as global functions, for example `pc.getValueAtPath` which is perhaps not ideal.

Maybe wrap those in a static class, for example like `ShaderUtils` and similar, to group them in a 'namespace'

Reply:
Well the Esm Script won't work in the ES5 bundle, so do we need actually these exported in the index?

Reply:
I think the module version of the engine is still based on index.js. @slimbuck ?

Reply:
Yes. External apps include public engine api from index.js. Otherwise we're exposing the internal file structure of the engine and apps break when we reorganise our file hierarchy.

Reply:
Gotya. ok well I might keep this internal for now. We can decide if and how to make this public when it becomes necessary

Reply:
This is now private